contrib/jms/src/main/java/com/google/cloud/dataflow/contrib/jms/JmsIO.java

```diff
@@ -0,0 +1,532 @@
+/*
+ * Copyright (C) 2015 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.dataflow.contrib.jms;
+
+import com.google.cloud.dataflow.sdk.coders.Coder;
+import com.google.cloud.dataflow.sdk.coders.StringUtf8Coder;
+import com.google.cloud.dataflow.sdk.coders.VoidCoder;
+import com.google.cloud.dataflow.sdk.transforms.Create;
+import com.google.cloud.dataflow.sdk.transforms.DoFn;
+import com.google.cloud.dataflow.sdk.transforms.PTransform;
+import com.google.cloud.dataflow.sdk.transforms.ParDo;
+import com.google.cloud.dataflow.sdk.util.CoderUtils;
+import com.google.cloud.dataflow.sdk.util.WindowingStrategy;
+import com.google.cloud.dataflow.sdk.values.PCollection;
+import com.google.cloud.dataflow.sdk.values.PDone;
+import com.google.cloud.dataflow.sdk.values.PInput;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.annotation.Nullable;
+
+import javax.jms.Connection;
+import javax.jms.ConnectionFactory;
+import javax.jms.MessageConsumer;
+import javax.jms.MessageProducer;
+import javax.jms.Session;
+import javax.jms.TextMessage;
+
+/**
+ * Read and Write {@link PTransform}s for JMS broker. These transforms create and consume
+ * unbounded {@link PCollection PCollections}.
+ */
+public class JmsIO {
+
+    private static final Logger LOG = LoggerFactory.getLogger(JmsIO.class);
+
+    /** The default {@link Coder} used to translate to/from JMS messages. */
+    public static final Coder<String> DEFAULT_JMS_CODER = StringUtf8Coder.of();
+
+    /**
+     * A {@link PTransform} that continuously reads from a JMS broker and
+     * return a {@link PCollection} of {@link String Strings} containing the items.
+     */
+    public static class Read {
+
+        /**
+         * Creates and returns a transform for reading from the JMS broker with the specified
+         * transform name.
+         *
+         * @param name the step name
+         * @return the Bound
+         */
+        public static Bound<String> named(String name) {
```

Comment:
per our discussion on the mailing list -- I would drop the `Bounded` class and instead just make `Read` itself be a `PTransform<PBegin, PCollection<T>>`. The name here is unnecessary -- it can be more easily set directly via `p.apply("name", JmsIO.Read...);`